src/main/scala/rise/core/traverse.scala

```diff
@@ -23,29 +24,29 @@ object traverse {
       for { t1 <- `type`(i.t)}
         yield i.setType(t1).asInstanceOf[I]
     def typeIdentifierDispatch[I <: Kind.Identifier] : VarType => I => M[I] = vt => i => (i match {
-      case n: NatIdentifier => bind(typeIdentifier(vt)(n))(nat)
-      case dt: DataTypeIdentifier => bind(typeIdentifier(vt)(dt))(datatype)
-      case a: AddressSpaceIdentifier => bind(typeIdentifier(vt)(a))(addressSpace)
-      case m: MatrixLayoutIdentifier => bind(typeIdentifier(vt)(m))(matrixLayout)
-      case f: FragmentKindIdentifier => bind(typeIdentifier(vt)(f))(fragmentKind)
-      case n2n: NatToNatIdentifier => bind(typeIdentifier(vt)(n2n))(natToNat)
-      case n2d: NatToDataIdentifier => bind(typeIdentifier(vt)(n2d))(natToData)
-      case t: TypeIdentifier => typeIdentifier(vt)(t)
+      case t: IType => typeIdentifier(vt)(t)
+      case n: INat => bind(bind(typeIdentifier(vt)(n))(i => nat(i.id)))(i => return_(INat(i.asInstanceOf[NatIdentifier])))
+      case dt: IDataType => bind(bind(typeIdentifier(vt)(dt))(i => datatype(i.id)))( i => return_(IDataType(i.asInstanceOf[DataTypeIdentifier])))
+      case a: IAddressSpace => bind(bind(typeIdentifier(vt)(a))(i => addressSpace(i.id)))( i => return_(IAddressSpace(i.asInstanceOf[AddressSpaceIdentifier])))
+      case m: IMatrixLayout => bind(bind(typeIdentifier(vt)(m))(i => matrixLayout(i.id)))( i => return_(IMatrixLayout(i.asInstanceOf[MatrixLayoutIdentifier])))
+      case f: IFragmentKind => bind(bind(typeIdentifier(vt)(f))(i => fragmentKind(i.id)))( i => return_(IFragmentKind(i.asInstanceOf[FragmentKindIdentifier])))
+      case n2n: INatToNat => bind(bind(typeIdentifier(vt)(n2n))(i => natToNat(i.id)))( i => return_(INatToNat(i.asInstanceOf[NatToNatIdentifier])))
+      case n2d: INatToData => bind(bind(typeIdentifier(vt)(n2d))(i => natToData(i.id)))( i => return_(INatToData(i.asInstanceOf[NatToDataIdentifier])))
```

Comment:
Why do we need casts here now?

Reply:
Before it was sufficient with the outer cast `.asInstanceOf[M[I]]`, now because we need to wrap it with the pertinent `Kind.Identifier` constructor we need to cast once in there, once outside (I should check if the outside one can be removed now though)

Reply:
> Looks good overall, although like for #181, I wish we could have less generic parameters. It feels like the kind should imply everything else (`T`, `I`, `KI`), not depend on it.

I totally agree, these type parameters leak out everywhere, it's pretty annoying (especially because they leak into some of the `Expr` constructors).

Reply:
Nope, Scala can't see that matching `I` against e.g. `INatToData` and determines `I` to be `INatToData` on the rhs of the case expression, and that therefore `return_(INatToData(_))` is ok. However what we can do is route through `nat`/`datatype`/etc first, then through `typeIdentifier`, that saves us from an extra bind, and the tests seem to pass :)